src/material/datepicker/testing/datepicker-trigger-harness-base.ts

```diff
@@ -0,0 +1,100 @@
+/**
+ * @license
+ * Copyright Google LLC All Rights Reserved.
+ *
+ * Use of this source code is governed by an MIT-style license that can be
+ * found in the LICENSE file at https://angular.io/license
+ */
+
+import {ComponentHarness, LocatorFactory, TestElement} from '@angular/cdk/testing';
+import {CalendarHarnessFilters} from './datepicker-harness-filters';
+import {MatCalendarHarness} from './calendar-harness';
+
+/** Interface for a test harness that can open and close a calendar. */
+export interface DatepickerTrigger {
```

Comment:
I'm not super happy with how this one turned out, but this is the best way I found to implement it. The problem is that these harnesses all have 2 base classes: an input and a datepicker trigger. The input is extended by `MatDatepickerInput`, `MatStartDate` and `MatEndDate`, whereas the trigger needs to be extended by `MatDatepickerInput`, `MatDateRangeInput` and `MatDatepickerToggle`. Initially I had the input base class extend the trigger, but the trigger behavior doesn't make sense on `MatStartDate` and `MatEndDate`. I tried to use a mixin so that `MatDatepickerInput` can extend both base classes, but TS kept complaining about the mixin class having the protected methods coming from `ComponentHarness`. 

I ended up with this which has an interface that enforces that `MatDatepickerInput` and the classes extending the trigger base class are all consistent while having the trickier shared logic in functions.

Reply:
I think that's totally reasonable